src/Npgsql/pg_catalog/pg_type.dat

```diff
@@ -0,0 +1,695 @@
+#----------------------------------------------------------------------
+#
+# pg_type.dat
+#    Initial contents of the pg_type system catalog.
+#
+# Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+# Portions Copyright (c) 1994, Regents of the University of California
+#
+# src/include/catalog/pg_type.dat
+#
+#----------------------------------------------------------------------
+
+[
+
+# For types used in the system catalogs, make sure the values here match
+# TypInfo[] in bootstrap.c.
+
+# OID symbol macro names for pg_type OIDs are not specified here because
+# they are generated by genbki.pl according to the following rule:
+#  foo_bar  ->  FOO_BAROID
+# _foo_bar  ->  FOO_BARARRAYOID
+
+# To autogenerate an array type, add 'array_type_oid => 'nnnn' to the element
+# type, which will instruct genbki.pl to generate a BKI entry for it.
+# In a few cases, the array type's properties don't match the normal pattern
+# so it can't be autogenerated; in such cases do not write array_type_oid.
+
+# Once upon a time these entries were ordered by OID.  Lately it's often
+# been the custom to insert new entries adjacent to related older entries.
+# Try to do one or the other though, don't just insert entries at random.
+
+# OIDS 1 - 99
+
+{ oid => '16', array_type_oid => '1000',
```

Comment:
We could set up a CI action which downloads the latest copy of this before building. Though manually updating from time to time should also be sufficient.

Reply:
> We could set up a CI action which downloads the latest copy of this before building.

You mean have them in our repo and make sure that they stay up to date with a CI action that runs nightly or so?
This would probably work but I'm not sure if it's worth it.

Reply:
The simple version would be to just download the file every time as part of the CI build - that doesn't commit it into source control, but ensures releases have the latest one. The version in source control may become stale, but that's not very important.

A more elaborate version would be a Github Actions workflow which periodically (weekly?) checks if the file changed, and sends a PR to merge the new one into the Npgsql repo. Definitely feels much for the task at hand - though could be nice for playing around with Github Actions if anyone's interested :)

But thinking about it some more... Maybe the thing to do would be to have a simple test which connects to a blank database with type loading, and then without type loading (your new mode), and asserts that all PG types in the former are also found in the latter (and even have the same OID). This way, if there's a change on the PG side, the test would fail and we'd know we need to update. What do you think?

Reply:
> But thinking about it some more... Maybe the thing to do would be to have a simple test which connects to a blank database with type loading, and then without type loading (your new mode), and asserts that all PG types in the former are also found in the latter (and even have the same OID). This way, if there's a change on the PG side, the test would fail and we'd know we need to update. What do you think?

The database to connect to would need to be at least the same major version as our .dat files (I've opted for the latest major release (14 right now) but not a development version), otherwise we might have additional types, but yes, that seems like a good Idea.
We'd have to advance the PG version we pin the test to whenever we upgrade the .dat files to a new catalog version but that seems like a minor problem, so yes - let's do that.

Reply:
Yep, makes sense.

If we do decide to have a Perl script generating JSON, it could integrate the PG version into that JSON, and the test could skip if the PG version doesn't match what it finds in the JSON.